### ISSUE TITLE
move section V5.5 (and V5.4) to V10

### ISSUE THREAD

**Issue by elarlang** [OWASP/ASVS#1903]
The topic was raised by Fabio Cerullo in Slack chat.

Current section [V5.5 Deserialization Prevention](https://github.com/OWASP/ASVS/blob/master/5.0/en/0x13-V5-Validation-Sanitization-Encoding.md#v55-deserialization-prevention) contains requirements which should be moved to "code review" category or section, we have V10 and V10.4 for those now.

The same applies to V5.4.

**Comment by tghosth** [comment]
I don't agree, I stand by my previous comments [in the thread](https://owasp.slack.com/archives/C06MNF14M/p1709595823045979):

> _deserialisation vulnerabilities are a result of not handling dangerous input correctly which is why they're there (**in 5.x**)_ 
>
> _ultimately this is a problem of bad input which is why I think v5 is the closest compared to other chapters_
>
>  I would add that I think we would rather have related requirements together rather than being too strict about which chapter they belong to

**Comment by jmanico** [comment]
I agree, Josh. But that is, of course, 1/2 that story.

Deserialization vulnerabilities occur when an application deserializes untrusted data without adequate validation or sanitization. An attacker can manipulate the serialized data to achieve arbitrary code execution, privilege escalation, or denial of service attacks.

While it is true that these vulnerabilities come from improperly handling input (which is often external), it's not just about "bad input" but also about how the application processes and uses this input. :D

**Comment by elarlang** [comment]
Just for brainstorm.

I think we can take a parallel here with the so-called XSS - input validation MAY give you a defensive effect, but the application needs to guarantee that it can process the data securely.

**Example 1**:

  * _V5.5.2 Verify that the application correctly restricts XML parsers to only use the most restrictive configuration possible and to ensure that unsafe features such as resolving external entities are disabled to prevent XML eXternal Entity (XXE) attacks._

You may validate the incoming input, but the application is still vulnerable when meeting malicious input. The requirement does not say anything about input validation, but strict configuration to avoid XXE to happen even when not validated input reach to the functionality. The requirement is correct, but the choice of section is questionable.

The input validation part for processing XML is covered by other requirement:

  * _V13.3.1 Verify that XSD schema validation takes place to ensure a properly formed XML document, followed by validation of each input field before any processing of that data takes place._


**Example 2**:

  * _V5.5.4 Verify that when parsing JSON in browsers or JavaScript-based backends, JSON.parse is used to parse the JSON document. Do not use eval() to parse JSON._

JSON.parse vs eval - you can not fix it with validation. Validation is business logic rule, if you have free text field, you can not disallow anything, you just can not use eval. It's about program code, not input validation.

**Comment by jmanico** [comment]
Elar: many forms of XSS defense MANDATES input validation.

1) User generated HTML is best defended against XSS via HTML validation 
on the client. See DomPurify.

2) User constructed URL's that land in active anchor tags are best 
defended against via URL validation in the client itself, like Angular 
has built in.

The whole notion that input validation is not crucial to XSS defense is 
simply not true, at all.

**Comment by elarlang** [comment]
Well, discussion over XSS is a bit offtopic here, but your example 1) is sanitization and 2) URL attribute value really needs validation.

We can spin this URL input validation with also - although you expect validated input, we still have Content-Security-Policy to mitigate loading external resources and requirement for accepting only allow-listed URL's to be loaded on the server side:

**Comment by tghosth** [comment]
I am open to a proposal, I would rather be fast rather than right in this case :)

**Comment by kwwall** [comment]
FWIW, here's my $.02. I agree with @tghosth here. I really don't think these deserialization-related requirements should be moved anymore than any of the other ASVS requirements in Chapter 5 should be moved. If you look at section v10.4, those are on "_general_ principles" supporting defensive coding, not specific techniques. If we move  V5.5 to v10.4, then we might as well move then entire Chapter 5 there because certainly input validation is a general defense and I think if anything fits under "defensive coding", that does. However, the techniques for deserialization defense are specific in nature. I know, as I've written a few in my day.

There are (at least) 2 distinct basic use cases that come into play here:
1. Verifying that serialized input came from a _trusted_ (based on the outcome of your threat model) source, implying that if you can confirm that the serialized input has a) not been tampered with and b) it actually originated from a trusted source (both of which you can do via cryptographic means like digital signatures, HMACs, or AEAD encryption) that you can use the serialized data "as-is" with otherwise validating the serialized data (although, I would contend it still is generally a good idea to do so, but that's something to decide in your threat model).
2. Verifying that serialized data that comes from a non-trusted source requires verifying that the data is "safe to use". What makes this tricky is that it may have to be verified in special ways because you can't deserialize it and check afterwards (does Chris Frohoff's [ysoserialize](https://github.com/frohoff/ysoserial) ring a bell?). In Java, you can white-list classes know to be "safe" (which is much trickier than it looks; I would not attempt it without some threat modeling) using something like [JEP 290](https://openjdk.org/jeps/290).

So, I don't think that these defensive techniques used to defend against deserialization vulnerabilities are general enough to move them to any code review section on defensive coding.

**Comment by jmanico** [comment]
I think this is a wise perspective, Daniel.

And I also want to state again that in general input validation is absolutely a critical defense later in many situations and we cannot have secure software without it. 😁

**Comment by elarlang** [comment]
I gave examples with requirements 5.5.2 and 5.5.4, but I _assume_ I got arguments against those with requirements 5.5.3. If this is the case we don't talk about the same thing and there is no chance to find agreement.

So, instead of describing abstract deserialization, I go requirement by requirement. Otherwise, there is a risk of taking one detail or requirement out of context and applying it to all requirements.


**5.5.2**

> Verify that the application correctly restricts XML parsers to only use the most restrictive configuration possible and to ensure that unsafe features such as resolving external entities are disabled to prevent XML eXternal Entity (XXE) attacks.

It's about configuration or setup for library. Nothing about processing user input. Fits to V10 or to some configuration hardening. Also fits to "general principle" category.

The requirement may need an exception "if not intended by the application for controlled input" or something.

Like pointed out previously - the input validation part for processing XML is covered by other requirement:

> V13.3.1 Verify that XSD schema validation takes place to ensure a properly formed XML document, followed by validation of each input field before any processing of that data takes place.


**5.5.3**

> Verify that deserialization is not used when communicating with untrusted clients. If this is not possible, ensure that deserialization is performed safely, for example, by only allowing a allow-list of object types or not allowing the client to define the object type to deserialize to, in order to prevent deserialization attacks.

This requirement needs some work, the construction "Don't use it! But if you do, do it that way." does not make sense as a requirement.

This requirement describes handling safely untrusted input when serializing, and fits to V5. I _assume_ this is also the requirement Kevin talked about and focused on.


**5.5.4**

> Verify that when parsing JSON in browsers or JavaScript-based backends, JSON.parse is used to parse the JSON document. Do not use eval() to parse JSON.

With or without untrusted input, eval() should never be used instead of JSON.parse(). It fits to V10.4. Also fits to "general principle" category.


**5.5.5**

> Verify that different parsers used in the application for the same data type (e.g. JSON parsers, XML parsers, URL parsers), perform parsing in a consistent way and use the same character encoding mechanism to avoid issues such as JSON Interoperability vulnerabilities or different URI or file parsing behavior being exploited in Remote File Inclusion (RFI) or Server-side Request Forgery (SSRF) attacks.

It is about architecture or choice of components to be used in the application. I don't think it belongs to V5. Not sure it belongs to V10.4 either. But fits to "general principle" category.


Comments to @kwwall  https://github.com/OWASP/ASVS/issues/1903#issuecomment-2015860600


> So, I don't think that these defensive techniques used to defend against deserialization vulnerabilities are general enough to move them to any code review section on defensive coding.

The conclusion from my text above is that there is a disagreement in 3 of 4 related requirements.


> Therefore, I vote to close this issue. Or, if you really wish to continue to discuss it, I'd propose promoting it to a Discussion.

FYI: for ASVS, we use issues to discuss certain categories or requirement-related problems or proposals, Discussions are used for more generic concepts. I'll close it when there is a decision and solution.

Was my assumption, that your feedback is mainly focused on requirement 5.5.3 correct?


Comments to @jmanico  https://github.com/OWASP/ASVS/issues/1903#issuecomment-2016414087

1) It was Kevin.

2) Stay on topic :)

**Comment by elarlang** [comment]
ping @kwwall - can we get your other $0.02? :)

**Comment by elarlang** [comment]
A proposal as a conclusion for my comment (https://github.com/OWASP/ASVS/issues/1903#issuecomment-2016445622):

  * move V5.5.2, V5.5.4 and V5.5.5 to section V10.4 Defensive Coding
  * improve the wording for V5.5.3
    * and need to think about the section/category, as it will be only one requirement for this section.

**Comment by kwwall** [comment]
Just now saw this email assigning me to the ticket. I don't recall seeing a
separate assignment email from GitHub though. (Or I saw it and forgot
it.)Shouldn't that have happened as well. (Lately have been focusing on
other things like moving my office to a bigger bedroom, doing taxes, and
preparing for a talk at our local OWASP chapter chapter meeting.) All those
things are behind me so I can take a deeper look. Or @elar can take it
instead.  Up to you.

-kevin

**Comment by tghosth** [comment]
> A proposal as a conclusion for my comment ([#1903 (comment)](https://github.com/OWASP/ASVS/issues/1903#issuecomment-2016445622)):
> 
> * move V5.5.2, V5.5.4 and V5.5.5 to section V10.4 Defensive Coding
> * improve the wording for V5.5.3    
>   * and need to think about the section/category, as it will be only one requirement for this section.

I am not 100% on board with this but have no strong objection

**Comment by kwwall** [comment]
I have some objections. I will provide details this evening (US/Eastern timezone).

**Comment by kwwall** [comment]
**TL;DR Warning**: The **BLUF** for this is -
1. I think there needs to be a _really_ high bar for making changes like these to ASVS 5.0. Minor improvements need not apply for section moving.
2. I think that section V10.4 Defensive Coding is a poor fit to move requirements 5.5.2, 5.5.4, and 5.5.5 and I will explain why. (And BTW, what is it with the ASVS' team's apparent obsession to prefix everything with a bloody 'V'? I think it's understood what we are referring to and it should only be used in cases where it's otherwise ambiguous. It especially annoys me when I see something like "Chapter V8 Data Protection". "Chapter V8"? Really? Great vegie drink (especially the Spicy V8), but a chapter name? WTF? But I digress.)

The rest of this treatise is a detailed explanation of those 2 points.


First off, I don't think that the OWASP ASVS team has--in the general sense--completely thought through the impact of moving / deleting / changing specific requirements. I look at ASVS as a "contract" of sorts (as in an implied programming contract that an SDK will remain stable and as much as possible stay backward compatible.

Now, one can argue based on semantic versioning, if you are going to change things around, you should do it now in 5.0 as going from 4.0.3 to 5.0 is a change in a "major" version allows you to do this. But just because you can, doesn't mean you should.

When ASVS encourages companies to customize ASVS to their own needs, that implies some stability. Consider a company using the current stable release, 4.0.3, uses the identify for (say) v4.0.3-5.5.2 and has that mentioned in several of their wiki pages and maybe in a several dozen Jira tickets that all are referencing (say) a vulnerability template for XXE.

Fast forward to a day after which v5.0 is released and that company starts to update it. Now ideally, the (relatively) easy thing is for them to search for all the "v.4.0.3-" and change them to "v5.0-" in their wiki pages, they'd be finished. But those using ASVS know it isn't going to be _that_ easy. We realize that some (okay, several) of the 4.0.3 requirements have moved / deleted / or rewritten so that then need to be (re)customized to fit that companies policies and standards. Fair enough. But let's try to minimize the churn here. Because if you don't, the truth is that companies will simply not migrate to 5.0. If will be just like when Apache Struts 2 came out and rather than moving from Apache Struts 1.x to 2.0, the looked for alternatives and most that I've seen end up abandoning Struts 2 in favor of Spring Framework. So hopefully, that can be a lesson-learned for all of us. If you make migration too difficult, people will either forgo updating or look for an alternative. And if that happens, that's a loss for everyone. So OWASP has to have a high bar before making changes to 5.0, and I for one don't think this qualifies. The following paragraphs explains why.


So @elarlang's proposal is:
> * move V5.5.2, V5.5.4 and V5.5.5 to section V10.4 Defensive Coding
> 
> * improve the wording for V5.5.3
>   
>   * and need to think about the section/category, as it will be only one requirement for this section.

I have no objection improving wording for 5.5.3. I do strongly object to moving 5.5.2, 5.5.3, and 5.5.5 to section V10.4 Defensive Coding.

Here are my reasons. There are several:
1. I'm not sure what section V10.4 is intended to be. It doesn't have a lead-in paragraph explaing it like almost all of our other sections do. So that means it means different things to different people. And perhaps that is part of the reason for the confusion here.
2. Given there is not descriptive paragraph for what section V10.4 intended to be, I take take my personal contextual clues for the chapter that it belongs to, which is Chapter V10 Malicious Code and Insecure Code. Hmmm. I have a problem with that as well.
3. IMO, "Malicious Code" should not confuse with the exploit to breach a vulnerability. I think the Control Objects for this chapter rule this out. So that leaves either a payload that is left behind by a successful breach or _intentionally_ malicious code that is planted by an insider or is the result of some upstream supply chain attack like those that have happened on pypi and Node repos. So I think that is a pretty narrow focus. (And I think that what's currently in section V10.2 reinforces that characterization. Context is king here.)
4. Do 5.5.2, 5.5.3, and 5.5.5 fit the "Insecure Code" part? It seemingly does. But so do about 90% of the other ASVS requirements, so in reality is doesn't because that category is too broad and allows Chapter V10 to become a dumping ground. And IMO, THAT'S. JUST. WRONG.
5. In https://github.com/OWASP/ASVS/issues/1903#issuecomment-2004740559 @elarlang commented that:
> Well, discussion over XSS is a bit offtopic here, but your example 1) is sanitization and 2) URL attribute value really needs validation.
I might remind you that @jmanico only brought this up because you first brought it up in https://github.com/OWASP/ASVS/issues/1903#issuecomment-2003387343. So I will defend Jim even if it can't remember my name and called me Daniel. :) But my point is, input validation _IS_ "defensive coding", so by your argument we should move all input validation related requirements to section V10.4. But we won't because it generally doesn't have much to do with intentionally malicious code (i.e., malware) and generally is ineffective against that.
6. In https://github.com/OWASP/ASVS/issues/1903#issuecomment-2016445622, @elarlang regarding 5.5.2, wrote:
> It's about configuration or setup for library. Nothing about processing user input. Fits to V10 or to some configuration hardening. Also fits to "general principle" category.
I completely and absolutely object to this proposition. First off, it is almost all making the right SDK calls to set options in XML related libraries and not just some configuration option that you set. (See the [OWASP XML External Entity Prevention Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/XML_External_Entity_Prevention_Cheat_Sheet.html) for examples of what I am referring to.) Secondly, you state that there is "nothing about processing user input". That's just not true. If the user attempts to attack your application via XXE is is the attacker's input that is in fact being examined and filtered by these XML libraries.

Anyhow, by now it is getting late and you're probably all bleary-eyed from reading this.

Just remember the next time you ask for "my other $.02", that my middle name is "TL;DR".

Regards.

**Comment by kwwall** [comment]
And just for the official record, for this particular issue, I vote that we **_close_** it with a "Won't fix" label (if you have such for ASVS issues). And if you want to reword V5.5.3 for clarification, create that as a separate issue if you need to track it.

**Comment by elarlang** [comment]
> I'm not sure what section V10.4 is intended to be.

"I'm not sure what it is but it is for sure not correct"?

Oeh. I expected some good argument-based feedback, but this is mostly just so off-topic. Basically not too much about the content of the issues or about the reasons to move them, just ranting about the "V" prefix and "don't move them, because I don't like moving things around".

Please, if you going to object something, first try to understand, what is going on in big picture and I'm happy to explain, what and why we have done with the structure.

V10 (<- yes, with the V prefix), is not fully re-organized yet and I agree it needs more work. We know it.

We don't move things around because we can, but because I / we are sure that those belong to better place.

To address the problem "I can not find the v4.0.3 requirement from the same place in the v5.0" I personally have spent a lot of time to provide mapping and list of changes.

> And just for the official record, for this particular issue, I vote that we close it with a "Won't fix" label (if you have such for ASVS issues). And if you want to reword V5.5.3 for clarification, create that as a separate issue if you need to track it.

We got your input, and in general, I already commented on that:
> FYI: for ASVS, we use issues to discuss certain categories or requirement-related problems or proposals, Discussions are used for more generic concepts. **I'll close it when there is a decision and solution.**

> And if you want to reword V5.5.3 for clarification, create that as a separate issue if you need to track it.

Separate is something to consider, when you want to address other topics than asked in the issue.

**Comment by tghosth** [comment]
5.5.3 - Suggestion from Josh and Elar.

> [MODIFIED, MERGED FROM 1.5.2] Verify that if deserialization is used when communicating with untrusted clients, the input is handled safely. For example, by only allowing a allow-list of object types or not allowing the client to define the object type to deserialize to, in order to prevent deserialization attacks.

I think that 5.4 should stay here.

Elar, to make decision about moving 5.5.

**Comment by elarlang** [comment]
Arguments about why to move them and why to not move them are written down. As it's not something urgent to do and nothing dramatic happens when it's not done, I close the issue for now. I'll revisit the topic after the V10 rewrite.
